[Console] fixed 10531

nacmartin opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #10531 |
| License | MIT |
| Doc PR | - |

This is a fix for #10531. It works by extracting all the parent namespaces of a command.

nacmartin committed:
[Console] fixed #10531

aitboudad commented on src/Symfony/Component/Console/Application.php:
```diff
@@ -469,10 +469,10 @@ public function getNamespaces()
         $namespaces = array();
         foreach ($this->commands as $command) {
-            $namespaces[] = $this->extractNamespace($command->getName());
+            $namespaces = array_merge($namespaces, $this->extractAllNamespaces($command->getName()));
```
array_merge is not necessary here
```php
$namespaces = $this->extractAllNamespaces($command->getName());
```

stof replied:
it is because of the loop

aitboudad replied:
right

stof commented on src/Symfony/Component/Console/Application.php:
```diff
@@ -1169,4 +1169,27 @@ private function splitStringByWidth($string, $width)
+     */
+    private function extractAllNamespaces($name)
+    {
+        $parts = explode(':', $name, -1);
```
just omit the third argument

nacmartin replied:
If we omit the `-1` in explode, it will return not only namespaces (`doctrine`, `doctrine:database`) but also the full command (`doctrine:database:create`). As this is also run for command aliases, it will return also full command aliases.

e.g. [this test](https://github.com/symfony/symfony/blob/c7bee8f861cc6f728e865f7e7b931c74a3302322/src/Symfony/Component/Console/Tests/ApplicationTest.php#L186) will fail, because instead of only `foo` as expected, `getNamespaces` will also return commands and aliases:

```
Array(
    0 => 'help'
    1 => 'list'
    2 => 'foo'
    3 => 'foo:bar'
    4 => 'afoobar'
    5 => 'foo:bar1'
    6 => 'afoobar1'
)
```

This third argument is equivalent to using `array_pop` after the explode, as the previously used function is doing: https://github.com/symfony/symfony/blob/c7bee8f861cc6f728e865f7e7b931c74a3302322/src/Symfony/Component/Console/Application.php#L1039

stof replied:
ah indeed. Can you add a comment in the code explaining that the -1 is here to skip the command short name when exploding ?

carlescliment replied:
A comment? Why not a meaningful method name?

nacmartin replied:
I have added a comment and squashed.

@carlescliment `extractNamespace` was extracting the full namespace of a command (`doctrine:database:create` -> `doctrine:database`) and `extractAllNamespaces` is extracting all the namespaces of a command, not just the full one (`doctrine` and `doctrine:database`). Do you have a better idea for the method name?

carlescliment replied:
`extractNamespace` is okay, but comments sometimes don't sync properly with the evolution of the code. My suggestion would be to apply extract method to the line containing the `explode()`, wich seems to need additional semantics, and name the method in consonance with the comment you just wrote. 

For example:

```
$parts = $this->explodeSkippingCommandName($name);
```

carlescliment replied:
No worries, the case we are looking at is not a big deal at all. Just wanted to point out that any comment can be turned into a method name.

:+1:

yguedidi commented on src/Symfony/Component/Console/Application.php:
```diff
@@ -1169,4 +1169,28 @@ private function splitStringByWidth($string, $width)
+
+        foreach ($parts as $part) {
+            if (count($namespaces)) {
+                $namespaces[] = end($namespaces).":$part";
```
Shouldn't this be `end($namespaces).':'.$part;`

stof replied:
we generally don't use interpolation in strings in Symfony (we may have a few places using it though). And we also use single quotes everywhere except when we explicitly need double quotes (for ``\n`` chars for instance)

nacmartin replied:
Ok. Thanks for pointing out and the clarification. I have commited the change.

nacmartin commented:
After the review I have added, as asked, [a comment](https://github.com/symfony/symfony/pull/13309/files#diff-38eebdda7d6ebfcb7405b91167f05c9cR1182), and changed double quotes for single quotes [here](https://github.com/symfony/symfony/pull/13309/files#diff-38eebdda7d6ebfcb7405b91167f05c9cR1188). Changes commited and squashed.

javiereguiluz commented:
:+1: nice work @nacmartin ... and thanks for the patience going through the commit review process!

stof commented:
OK, :+1:

fabpot commented:
Thank you @nacmartin.
